gap/PolygonalComplexes/modification.gi

```diff
@@ -2046,6 +2046,171 @@ InstallMethod( ButterflyInsertion,
 end
 );
 
+BindGlobal("__SIMPLICIAL_DoButterflyDeletion",
+function(surface, f1, f2, edgesOfF1, edgesOfF2, intersectingEdges)
+    local joinPair, intersectingEdge, edgePath1, edgePath2,
+        outerFaceVertices, innerFaceVertices, reduceEdgePaths;
+
+    if Length(intersectingEdges) = 2 then
+        surface := RemoveFacesNC(surface, [f1, f2]);
+
+        edgesOfF1 := Filtered(edgesOfF1, x -> not x in intersectingEdges);
+        edgesOfF2 := Filtered(edgesOfF2, x -> not x in intersectingEdges);
+
+        joinPair := JoinEdges(surface, edgesOfF1[1], edgesOfF2[1]);
+
+        surface := joinPair[1];
+
+        return [surface, VertexEdgePathByEdges(surface, [joinPair[2]])];
+    elif Length(intersectingEdges) = 3 then
+        return [RemoveFacesNC(surface, [f1, f2]), ];
+    fi;
+
+    intersectingEdge := intersectingEdges[1];
+
+    # Collect inner face vertices
+    innerFaceVertices := VerticesOfEdge(surface, intersectingEdge);
+
+    # Collect outer face vertices
+    outerFaceVertices := [
+        Difference(VerticesOfFace(surface, f1), innerFaceVertices)[1],
+        Difference(VerticesOfFace(surface, f2), innerFaceVertices)[1]
+    ];
+
+    # Explicitly define the edge paths
+    edgePath1 := [outerFaceVertices[1], innerFaceVertices[1], outerFaceVertices[2]];
+    edgePath2 := [outerFaceVertices[1], innerFaceVertices[2], outerFaceVertices[2]];
+    
+    # Check if at least one of the edge paths is on the border of surface
+    if Length(FacesOfVertex(surface, innerFaceVertices[1])) = 2 or
+       Length(FacesOfVertex(surface, innerFaceVertices[2])) = 2 then
+        # Since at least one of the edge paths is on the border of surface we can just
+        # remove the faces f1 and f2 from surface and we dont have to join edge paths
+
+        surface := RemoveFacesNC(surface, [f1, f2]);
+
+        return [surface, ];
+    fi;
+    
+    reduceEdgePaths := function(surface, face, edgePathPos)
+        local faceVertex;
+
+        if Length(NeighbourFacesOfFace(surface, face)) = 3 then
+            return;
+        fi;
+        # Since face does not have two neighbours to merge the edge path, we
+        # remove the vertex from both edge paths that is unique to face
+
+        Remove(edgePath1, edgePathPos);
+        Remove(edgePath2, edgePathPos);
+    end;
+
+    reduceEdgePaths(surface, f1, 1);
+    reduceEdgePaths(surface, f2, 3);
```

Comment:
I think it is easier to read and a better style without this extra function `reduceEdgePaths` since nothing significant happens here.